packages/react-native/ReactCommon/react/renderer/components/view/BaseViewProps.cpp

```diff
@@ -392,19 +392,33 @@ Transform BaseViewProps::resolveTransform(
     LayoutMetrics const& layoutMetrics) const {
   float viewWidth = layoutMetrics.frame.size.width;
   float viewHeight = layoutMetrics.frame.size.height;
-  if (!transformOrigin.isSet() || (viewWidth == 0 && viewHeight == 0)) {
-    return transform;
+  auto transformMatrix = Transform{};
+
+  if (viewWidth == 0 && viewHeight == 0){
+    return transformMatrix;
   }
-  std::array<float, 3> translateOffsets =
-      getTranslateForTransformOrigin(viewWidth, viewHeight, transformOrigin);
-  auto newTransform = Transform::Translate(
-      translateOffsets[0], translateOffsets[1], translateOffsets[2]);
-  newTransform = newTransform * transform;
-  newTransform =
-      newTransform *
+
+  // transform is matrix
+  if (transform.operations.size() == 1 && transform.operations[0].type == TransformOperationType::Arbitrary) {
+    transformMatrix = transformMatrix * transform;
+  } else {
+    for (size_t i = 0; i < transform.operations.size(); i++) {
+      auto transformFromOperation = transform.FromTransformOperation(transform.operations[i], viewWidth, viewHeight);
+      transformMatrix = transformMatrix * transformFromOperation;
+    }
+  }
+      
+  if (transformOrigin.isSet()) {
+    std::array<float, 3> translateOffsets =
+          getTranslateForTransformOrigin(viewWidth, viewHeight, transformOrigin);
+    transformMatrix = 
       Transform::Translate(
-          -translateOffsets[0], -translateOffsets[1], -translateOffsets[2]);
-  return newTransform;
+            translateOffsets[0], translateOffsets[1], translateOffsets[2])
+      * transformMatrix
+      * Transform::Translate(-translateOffsets[0], -translateOffsets[1], -translateOffsets[2]);
+  }
```

Comment:
Why not do this before the transform processing loop? That way we can avoid less intermediate objects.

Reply:
We can do `translateOffsets[0], translateOffsets[1], translateOffsets[2])` before the loop but `-translateOffsets[0], -translateOffsets[1], -translateOffsets[2]` needs to be done post processing for transform origin to work correctly. I am missing how it can lead to less intermediate objects. Let me know, happy to fix it! Also, we want to prevent additional `transformOrigin.isSet()` checks.